### ISSUE TITLE
Troubles in predicting pose

### ISSUE THREAD

**Issue by GiBucci** [wangsen1312/unsupervised3dhuman#4]
Hi.

I'd like to start thanking you, because I was looking for solutions on how to fit a point cloud for animation reasons, and your work is the best I could find on the topic, it is very close to what I'm trying to do and you published some code, so thank you so much for your effort.

Sorry to disturb you, but I was experimenting with your algorithm on body scans, and I came across some issues.
I am talking about body facing direction, legs crossing and body parts compenetration.

I see that in previous issues you already solved these issues, but it is not working for me.

I added the row -- pred_pose[0, 1] = (pred_pose[0, 1] + 1.57).unsqueeze(0).float() after line 100, but it is still not working
I suspected the reason could be in the fact that my body is oriented along a different axis, but the model actually predicts the orientation correctly, it just places the body rotated 180 degrees (face towards the back of the original point cloud). This creates a pose estimation that is completely wrong.
I am working on point clouds of various people in a slight A-pose, so I guess it should be a very easy task, so it is the only explanation for those weird poses generated.

Any suggestion for this? Should I apply rotations to my original point cloud in advance? Am I doing something wrong? I was trying to make your code work before trying and training my own solution.

About leg crossing, I saw you applied another correction, but the issue is in a language I don't understand, so it is difficult for me to understand if it has been already patched, or where should I change the code.

I don't know if you are still working on this, and if you can help me, but I would be very grateful if you could find a minute to drop me some hints

Thank you in advance for your consideration and your help 

Have a nice day

**Comment by wangsen1312** [comment]
@GiBucci Sorry for the late reply. 
I suggest you do the rotation (you can rotate to point cloud face to the front if you can) before the optimization. Since our code is based on point cloud and we don't use the normal information, the front/ back problem will exists.

**Comment by wangsen1312** [comment]
@GiBucci For the leg problem 
The user add the leg_preserve_loss as below:
leg_preserve_loss = (10 ** 6) * ((body_pose[:, 2] - 0.0) ** 2).sum(dim=-1) +
(10 ** 6) * ((body_pose[:, 5] + 0.0) ** 2).sum(dim=-1)

total_loss = correspond_loss + pose_prior_loss + shape_prior_loss + betas_preserve_loss +
pose_preserve_loss + chamfer_loss + point2mesh_loss + leg_preserve_loss

The assumption is the leg didn't cross too much. You can add you want.

**Comment by GiBucci** [comment]
Hi @wangsen1312,

thank you so much for your answer! It is appreciated anyway.

I already solved the rotation problem, and I will surely test the new loss you suggested. In my case the poses are quite standardized so I don't expect much trouble in that.

The only problem that is convincing me to find another way is the fitting of the point cloud.
It seems like, while the pose is predicted quite correctly, the shape isn't "inflated" (or reduced) enough, leaving me with a model that is very close to the standardized model, and not even close to the point cloud I gave him as an input.
This is especially true for prominent parts, like abdomen, female chest, and so on.

Even after reading your code I still quite don't understand what to tune in order to tell the model to be more "fitting", can you give me a hint on what could be changed? Or am I doing something wrong?

I tried playing with the losses and the number of iterations, but nothing seems to be working.

Just so you know, I'm still working with SMPL and FAUST dataset for testing purposes, so I don't think it's a dataset problem since it's the same that you use.

Thank you again for your time, and hope you have a nice day.

**Comment by wangsen1312** [comment]
@GiBucci  If you would like the shape to be more "fitting", I would like you adjust the  shape_prior_loss , the shape_prior weight can be changed smaller than what we used now.  Hope this can help you.

**Comment by GiBucci** [comment]
Hi @wangsen1312, thank you for your reply.

I brought the shape_prior weight to zero, and now it finally works better. The only problem is that is still not quite precise when it comes to the belly and the female chest. I am trying again playing with weights and interactions, if you have any other tip I would be glad to hear them.

In any case thank you again.

**Comment by wangsen1312** [comment]
@GiBucci  Since our system rely on prior to make all the things working properly.
I have one suggestion: you can use SMPL + offset to optimize it.  Since the current results is very close to the real results, you should remember to regularize the offset to make it small enough. Try to use small learning rate, get each step visual results.
